Zend/zend_fibers.c

```diff
@@ -0,0 +1,778 @@
+}
+/* }}} */
+
+/* {{{ proto mixed Fiber::start(mixed ...$args) */
+ZEND_METHOD(Fiber, start)
+{
+	zend_fiber *fiber = (zend_fiber *) Z_OBJ_P(getThis());
+
+	if (fiber->status != ZEND_FIBER_STATUS_INIT) {
+		zend_throw_error(zend_ce_fiber_error, "Cannot start a fiber that has already been started");
+		return;
+	}
+
+	fiber->fci.params = ZEND_CALL_ARG(execute_data, 1);
+	fiber->fci.param_count = ZEND_CALL_NUM_ARGS(execute_data);
+
+	if (!zend_fiber_init_context(&fiber->context, zend_fiber_execute, EG(fiber_stack_size))) {
+		zend_throw_error(NULL, "Could not create fiber context");
+		return;
+	}
+
+	fiber->status = ZEND_FIBER_STATUS_RUNNING;
+
+	zend_fiber_switch_to(fiber);
+
+	zval_ptr_dtor(&fiber->fci.function_name);
+
+	if (fiber->status & ZEND_FIBER_STATUS_FINISHED) {
+		RETURN_NULL();
+	}
+
+	RETVAL_COPY_VALUE(&fiber->value);
+	ZVAL_UNDEF(&fiber->value);
+}
+/* }}} */
+
+/* {{{ proto mixed Fiber::suspend(mixed $value) */
+ZEND_METHOD(Fiber, suspend)
+{
+	zend_fiber *fiber = EG(current_fiber);
+	zval *error, *value = NULL;
+
+	if (UNEXPECTED(!fiber)) {
+		zend_throw_error(zend_ce_fiber_error, "Cannot suspend outside of a fiber");
+		return;
+	}
+
+	if (UNEXPECTED(fiber->status != ZEND_FIBER_STATUS_RUNNING)) {
+		if (fiber->status == ZEND_FIBER_STATUS_SHUTDOWN) {
+			zend_throw_error(zend_ce_fiber_error, "Cannot suspend in a force closed fiber");
+		} else {
+			zend_throw_error(zend_ce_fiber_error, "Cannot suspend in a fiber that is not running");
+		}
+		return;
+	}
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_ZVAL(value);
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (value) {
+		ZVAL_COPY(&fiber->value, value);
+	} else {
+		ZVAL_NULL(&fiber->value);
+	}
+
+	fiber->execute_data = execute_data;
+	fiber->status = ZEND_FIBER_STATUS_SUSPENDED;
+
+	// Remove running reference while suspended.
+	GC_DELREF(&fiber->std);
+
+	zend_fiber_suspend(fiber);
+
+	if (fiber->status == ZEND_FIBER_STATUS_SHUTDOWN) {
+		// This occurs when the fiber is GC'ed while suspended, do not add a ref.
+		if (EG(fiber_error)) {
+			zend_throw_unwind_exit();
+		} else {
+			zend_throw_fiber_exit();
+		}
+		return;
+	}
+
+	fiber->status = ZEND_FIBER_STATUS_RUNNING;
+	// Add reference while fiber is running.
+	GC_ADDREF(&fiber->std);
+
+	if (!fiber->error) {
+		RETVAL_COPY_VALUE(&fiber->value);
+		ZVAL_UNDEF(&fiber->value);
+		return;
+	}
+
+	error = fiber->error;
+	fiber->error = NULL;
+
+	execute_data->opline--;
+	zend_throw_exception_object(error);
+	execute_data->opline++;
+}
+/* }}} */
+
+/* {{{ proto mixed Fiber::resume(mixed $value = null) */
+ZEND_METHOD(Fiber, resume)
+{
+	zend_fiber *fiber;
+	zval *value = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_ZVAL(value);
+	ZEND_PARSE_PARAMETERS_END();
+
+	fiber = (zend_fiber *) Z_OBJ_P(getThis());
+
+	if (UNEXPECTED(fiber->status != ZEND_FIBER_STATUS_SUSPENDED)) {
+		zend_throw_error(zend_ce_fiber_error, "Cannot resume a fiber that is not suspended");
+		return;
```

Comment:
```suggestion
		RETURN_THROWS();
```
(also in various other methods)